[크롱] 뷰 챌린지 미션 1단계 제출합니다.

krrong opened the pull request:
안녕하세요 하티! 만나서 반갑습니다 🤠  
View의 라이프사이클을 같이 공부한게 엊그제 같은데 벌써 레벨4네요.. 시간이 너무 빠른 것 같아요ㅠ.ㅠ 
즐거운 추석을 보내기 위해 미션을 조금 빨리 진행했으니 부담갖지 마시고 리뷰는 천천히 부탁드릴게요!! 🙃  

## 구현 설명
리뷰 편의를 위해 구현에 대해 간단한 설명을 하도록 하겠습니다!

```kotlin
data class Point(
    val x: Float,
    val y: Float,
    val isStart: Boolean,
    val strokeWidth: Float,
    val color: Int,
)
```
점을 의미하는 Point 클래스를 만들어줬어요! 이 점을 그릴 때의 좌표 x, y와 선의 굵기(strokeWidth), 선의 색상(color)를 함께 가지고 있습니다.  
또 이 점이 시작지점인가도 함께 저장해요. 이렇게 구현한 이유는 아래에서 설명드릴게요 :)

```kotlin
    private fun drawPoint(canvas: Canvas, point: Point) {
        canvas.drawPoint(point.x, point.y, paint)
    }

    private fun drawLine(canvas: Canvas, startPoint: Point, endPoint: Point) {
        canvas.drawLine(startPoint.x, startPoint.y, endPoint.x, endPoint.y, paint)
    }
```
이렇게 두 개의 함수를 통해 onDraw() 메서드를 실행하는데요, 화면을 한 번 터치한 경우는 Point를, 화면에서 드래그를 한 경우 Line을 그리도록 했습니다!  
drawLine 메서드는 이전 Point와 현재 Point를 가져와서 선분을 잇는 형식인데요, Point가 시작점일 때 다른 처리를 하도록 해주지 않으면 이전 선분의 마지막 지점과 새로 긋는 선분의 첫 지점이 이어지는 현상이 발생합니다!  
그래서 Point에 isStart를 추가해주었어요!  

혹시나 이해되지 않는 부분이 있으면 언제든지 연락주셔도 됩니다!  
꼼꼼한 리뷰 부탁드릴게요 🙇‍♂️
좋은 하루 보내세요!

krrong committed:
docs: README 파일 작성

krrong committed:
feat: 자유 곡선을 그리는 기능 구현

krrong committed:
feat: viewBinding 추가

krrong committed:
feat: 도구 패널 뷰 추가

krrong committed:
feat: 브러시 굵기 변경기능 구현

krrong committed:
feat: 브러시 색상을 변경기능 구현

krrong committed:
feat: Point 클래스 생성

krrong committed:
feat: drawPath에서 drawLine로 수정

krrong committed:
refactor: 불필요한 로그 삭제

krrong committed:
refactor: rangebar 범위 설정

krrong committed:
refactor: 시작점임을 표시하는 변수 추가

krrong committed:
refactor: 시작점과 시작점이 아님에 따라 다르게 처리

- 시작점인 경우 point를 그리도록 함
- 시작점이 아닌 경우 line을 그리도록 함

sujin9 commented on app/src/main/java/woowacourse/paint/custom/CustomView.kt:
```diff
@@ -0,0 +1,69 @@
+
+    init {
+        isFocusable = true
+        isFocusableInTouchMode = true
```
이 두 줄의 역할은 무엇인가요?
둘의 역할이 무엇인지, 그리고 차이점이 무엇인지 궁금해서 질문 남깁니다 :)

krrong replied:
isFocusable는 해당 뷰가 포커스를 받을 수 있는지를 나타내고 true로 설정하면 뷰가 포커스를 받을 수 있습니다.  
isFocusableInTouchMode는 터치 모드에서 해당 뷰가 포커스를 받을 수 있는지 여부를 나타냅니다.  

사용한 이유는 강의자료에 있기 때문에 어딘가에 필요하겠지~ 하고 그대로 가져왔었는데요. 사실 이 코드가 없어도 정상적으로 동작이 됩니다.  
그리고 제 예상으로는 editText와 같이 포커스를 받을 수 있는 뷰와 같이 있을 때 언급하신 두 줄의 코드가 없으면 문제가 되겠다 싶었는데 확인해보니 그것도 아니더라구요...^^  
정의는 알겠으나 어떻게 사용해야 하는가는 아직 의문입니다..🤔

sujin9 replied:
저도 찾아보아도 잘 모르겠어서 질문을 드렸어요! 답변 감사합니다 :)

질문을 드리고 저도 추가적으로 찾아보았는데요,
아마 `접근성`과 관련이 있는 것 같아요.
예를 들면, talkback의 경우 focus가 있는 뷰를 읽어주는 것 같더라구요! 

ref.
https://developer.android.com/guide/topics/ui/accessibility/principles?hl=ko
https://gamjatwigim.tistory.com/200 
https://techblog.woowahan.com/2660/

sujin9 commented on app/src/main/res/layout/activity_main.xml:
```diff
@@ -6,4 +6,86 @@
+
+    <Button
+        android:id="@+id/btn_orange"
+        android:layout_width="0dp"
+        android:layout_height="wrap_content"
+        android:layout_marginStart="16dp"
+        android:backgroundTint="@color/orange"
+        app:layout_constraintBottom_toBottomOf="@id/btn_red"
+        app:layout_constraintEnd_toStartOf="@id/btn_yellow"
+        app:layout_constraintStart_toEndOf="@id/btn_red" />
+
+    <Button
+        android:id="@+id/btn_yellow"
+        android:layout_width="0dp"
+        android:layout_height="wrap_content"
+        android:layout_marginStart="16dp"
+        android:backgroundTint="@color/yellow"
+        app:layout_constraintBottom_toBottomOf="@id/btn_red"
+        app:layout_constraintEnd_toStartOf="@id/btn_green"
+        app:layout_constraintStart_toEndOf="@id/btn_orange" />
+
+    <Button
+        android:id="@+id/btn_green"
+        android:layout_width="0dp"
+        android:layout_height="wrap_content"
+        android:layout_marginStart="16dp"
+        android:backgroundTint="@color/green"
+        app:layout_constraintBottom_toBottomOf="@id/btn_red"
+        app:layout_constraintEnd_toStartOf="@id/btn_blue"
+        app:layout_constraintStart_toEndOf="@id/btn_yellow" />
+
+    <Button
+        android:id="@+id/btn_blue"
+        android:layout_width="0dp"
+        android:layout_height="wrap_content"
+        android:layout_marginStart="16dp"
+        android:backgroundTint="@color/blue"
+        app:layout_constraintBottom_toBottomOf="@id/btn_red"
+        app:layout_constraintEnd_toEndOf="@id/btn_thickness_change"
+        app:layout_constraintStart_toEndOf="@id/btn_green" />
```
하나하나 정성스럽게 추가해주셨네요..! 😮 👍
다만, 색상이 추가되면 코드도 계속 길어질텐데, 
이러한 확장성을 고려하여 개선해보면 어떨까요?

krrong replied:
ConstraintLayout에서 뷰에 Chain을 걸어서 할 수 있겠다! 라고 생각하고 이렇게 구현했어요..ㅎㅎ  
하티 의견대로 색상이 추가되면 xml코드가 길어지고 복잡해질 것 같아요!  
RecyclerView를 사용하여 반영하였습니다!

sujin9 commented on app/src/main/java/woowacourse/paint/MainActivity.kt:
```diff
@@ -1,14 +1,51 @@
+        viewBinding = ActivityMainBinding.inflate(layoutInflater)
+        setContentView(viewBinding.root)
+        setListener()
+        viewBinding.rangebar.valueFrom = 0.0f
+        viewBinding.rangebar.valueTo = 100.0f
```
```suggestion
android:valueFrom="0"
android:valueTo="100"
```
이 부분은 XML에서도 설정이 가능해요 ;)

krrong replied:
그런 방법도 있었군요! 감사합니다 🙃  
반영했고, 추가로 아래와 같은 속성도 있더군요!
```
android:stepSize="5"
```
이 속성을 설정해주면 Slider의 Step을 몇으로 할지 정할 수 있답니다!  
https://developer.android.com/reference/com/google/android/material/slider/RangeSlider#attr_Slider_android_stepSize

sujin9 replied:
오,, 이런 속성은 몰랐네요!! 감사합니다 :)
이번 미션의 경우로 예를 들면, brush의 두께 크기를 일정한 간격의 단위로 정한 경우 유용한 설정이겠네요 ;)

sujin9 commented on app/src/main/java/woowacourse/paint/MainActivity.kt:
```diff
@@ -1,14 +1,51 @@
+        viewBinding.rangebar.addOnChangeListener { _, value, _ ->
+            thickness = value
+        }
+        viewBinding.btnRed.setOnClickListener(this)
+        viewBinding.btnOrange.setOnClickListener(this)
+        viewBinding.btnYellow.setOnClickListener(this)
+        viewBinding.btnGreen.setOnClickListener(this)
+        viewBinding.btnBlue.setOnClickListener(this)
+    }
 
+    override fun onClick(v: View) {
+        when (v.id) {
+            R.id.btn_red -> color = getColor(R.color.red)
+            R.id.btn_orange -> color = getColor(R.color.orange)
+            R.id.btn_yellow -> color = getColor(R.color.yellow)
+            R.id.btn_green -> color = getColor(R.color.green)
+            R.id.btn_blue -> color = getColor(R.color.blue)
+        }
```
경우의 수 각각을 직접 나열하고 있는데, 
이 부분 역시 추후에 색상이 많아지면 코드가 너무 길어질 것 같아요!
깔꼼하게 단일화(?)해서 표현할 수 있도록 개선해보면 어떨까요?

krrong replied:
위에서 언급한대로 RecyclerView를 사용하여 반영하였습니다! 🙃

sujin9 commented on app/src/main/java/woowacourse/paint/custom/CustomView.kt:
```diff
@@ -0,0 +1,69 @@
+        super.onDraw(canvas)
+
+        val size = points.size
+        for (index in 1 until size) {
```
index를 0이 아닌 1부터 시작한 이유는 무엇인가요? 🤔

krrong replied:
하나의 선을 그리고 난 뒤에 다른 선분을 그리게 되면 두 선분의 끝점과 시작점이 이어지는 문제가 있었는데요, 이 것을 해결하려고 하다 미쳐 확인하지 못한 것 같아요 :0  
짚어주셔서 감사합니다 🙇‍♂️

sujin9 commented on app/src/main/java/woowacourse/paint/custom/CustomView.kt:
```diff
@@ -0,0 +1,69 @@
+        isFocusableInTouchMode = true
+    }
+
+    override fun onDraw(canvas: Canvas) {
```
수업 시간에 나왔던 `Path`와 `canvas.drawPath(Path, Paint)` 함수를 사용하지 않고
터치한 점들을 기록해서 점 또는 선을 그려주는 함수를 직접 구현해주셨네요! 👍 

다만, 선을 그어보면 마치 크레파스로 그림을 그리는 것처럼 삐뚤빼뚤하고 선이 매끄럽지 않아요. 
이는 자유 곡선처럼 보이지 않고, 직선들이 이어져있는 것처럼 보이고 있어요.
물론 그림을 그리는 듯한 느낌인 지금도 매력있지만,
요구사항에 맞게 좀 더 곡선답게 보일 수 있도록 이를 개선해보아요! :)

또한, onDraw에서 수행하는 작업이 많은 것 같아요.
저도 지식이 깊지 않아 정확하지 않을 수 있고, 체감이 잘 되지 않을 수도 있어요.
다만, 현재 코드를 보면, 그림을 그릴수록 point의 개수는 계속해서 늘어나고, 
onDraw가 호출될 때마다 이 point 개수만큼 반복문이 돌고, 이 반복문 안에서는 또 조건문이 있어요.
`onDraw`에서 수행하는 작업이 간결해지도록 개선해보면 좋겠어요! :)

krrong replied:
> 다만, 선을 그어보면 마치 크레파스로 그림을 그리는 것처럼 삐뚤빼뚤하고 선이 매끄럽지 않아요.
이는 자유 곡선처럼 보이지 않고, 직선들이 이어져있는 것처럼 보이고 있어요.
요구사항에 맞게 이를 개선해보아요! :)

<img src="https://github.com/woowacourse/android-paint/assets/84285337/af3c9e66-bb28-40cd-8343-55cf01861667" width = 180 />   

Paint에 strokeJoin, strokeCap 속성을 변경하여 매끄럽게 이어지도록 변경했습니다!
https://developer.android.com/reference/kotlin/androidx/compose/ui/graphics/StrokeJoin
https://developer.android.com/reference/kotlin/androidx/compose/ui/graphics/StrokeCap

> 또한, onDraw에서 수행하는 작업이 많은 것 같아요.

하티의 의견에 동의합니다!  
그래서 path를 사용하는 방법으로 변경했습니다 🤠  
저도 onDraw에서 하는 작업이 많아 무거워질 수 있다는 생각을 동일하게 하고 있었어요. 그런데 여기서 든 의문은 어디부터 무거운 작업으로 볼 것이냐에요. 다른 방법(path를 사용)을 사용한다고 하더라도 이전에 그려져있던 그림들을 다시 그리는 작업은 필수라고 생각하는데(반복문이 필요한지는 잘 모르겠지만요.) 어디부터가 무거운 작업에 속할까요?  
하티는 어떻게 생각하시나요?🙃

sujin9 replied:
> 어디부터가 무거운 작업에 속할까요?

저도 아직 무엇이 무거운 작업인가-에 명확한 기준을 찾지는 못한 것 같아요. 

다만, 기존 코드를 비교해서 얘기해보자면, 
(1) Path들에 대한 반복문이 아닌 Point에 대한 반복문이라면 반복을 도는 횟수도 훨씬 커지게 되고,
(2) 반복문에서 단순 그리는 작업 뿐만이 아니라 paint 색상, 두께를 지정하는 작업도 하고 있었고,  점을 찍을지 선을 그릴지를 구분하는 작업도 하고 있었어요.
위에서 언급한 작업들의 무게는 그다지 무겁지 않을 수도 있고, 큰 차이가 없을 지도 모르겠어요.
다만, 작업량이 더 작지는 않을 것이라고 생각하여 리뷰드렸던 거였어요! ;)

혹시 크롱께서는 어디부터가 무거운 작업일지 생각하시는 기준이 있으실까요?

sujin9 replied:
근데 크롱 hoxy ,,

> https://developer.android.com/reference/kotlin/androidx/compose/ui/graphics/StrokeJoin
https://developer.android.com/reference/kotlin/androidx/compose/ui/graphics/StrokeCap

compose 공부 중이신가요? 😆 😆 

는 농담이고,

https://developer.android.com/reference/android/graphics/Paint.Join
https://developer.android.com/reference/android/graphics/Paint.Cap

compose긴 해도 기존 문서와 다른 점이 크게 없는 것 같긴 하네요! 😆

sujin9 commented on app/src/main/java/woowacourse/paint/custom/CustomView.kt:
```diff
@@ -0,0 +1,69 @@
+
+        val size = points.size
+        for (index in 1 until size) {
+            paint.color = points[index].color
+            paint.strokeWidth = points[index].strokeWidth
```
하나의 선은 같은 색과 두께를 갖고 있을텐데,
모든 점에 대해 매번 색과 두께를 지정해줄 필요가 있을까요? 🤔

krrong replied:
점이 이어져 선분이 되는 것이기 때문에 점의 굵기와 색상을 점마다 가지고 있어야 한다고 생각했는데, 하티의 말을 듣고보니 점들의 집합(하나의 선)에 대해 색상과 두께를 지정해주면 더 효율적일 수 있겠다는 생각이 들어요!👍  
현재는 점들로 선을 이어주는 방식에서 다른 방식으로 변경했습니다!🦖

sujin9 commented on app/src/main/java/woowacourse/paint/custom/Point.kt:
```diff
@@ -0,0 +1,9 @@
+    val y: Float,
+    val isStart: Boolean,
+    val strokeWidth: Float,
+    val color: Int,
```
`@ColorInt` 어노테이션을 활용해보면 어떨까요?
(꼭 이 부분만이 아닌 코드 전반에서요!)

- 해당 어노테이션을 추가하면 어떤 이점이 있을까요?
- `@ColorInt` 와 `@ColorRes` 의 차이는 무엇일까요?

krrong replied:
`@ColorInt`
- 해당 매개변수나 리턴 값이 **색상 값**으로 해석되어야 함을 나타냅니다.
- Color.parseColor("#RRGGBB")와 같은 메서드에서 사용합니다.

`@ColorRes`
- 해당 매개변수나 리턴 값이 **색상 리소스 ID**로 해석되어야 함을 나타내는 애노테이션입니다.
- R.color.red와 같은 리소스 ID를 사용하여 색상을 지정할 때 사용합니다.
- 리소스 ID는 int이며, res/values/colors.xml과 같은 리소스 파일에 정의되어 있는 색상을 뜻합니다.

위와 같은 애노테이션을 추가함으로써 어떤 값이 들어와야 하는지 더 명확하게 구분할 수 있겠네요! 덕분에 좋은 정보 알아갑니다 감사해요!🦖

sujin9 replied:
네 맞습니다!! 😄 꼼꼼히 잘 찾아주셨네요 ~ 👍 

어떤 값이 들어올 지가 구분되는 가독성 측면의 장점도 있고,
IDE 상에서 올바르지 않은 리소스 타입이 전달되면 에러 (혹은 경고) 표시를 보여주어 실수를 줄일 수 있답니다! 😆 

ref.
https://developer.android.com/studio/write/annotations?hl=ko#res-annotations

sujin9 commented on app/src/main/res/layout/activity_main.xml:
```diff
@@ -6,4 +6,86 @@
+
+    <com.google.android.material.slider.RangeSlider
+        android:id="@+id/rangebar"
+        android:layout_width="match_parent"
```
```suggestion
        android:layout_width="0dp"
```
ConstraintLayout의 자식뷰는 match_parent 대신 0dp 사용을 권장해요!
https://developer.android.com/develop/ui/views/layout/constraint-layout#adjust-the-view-size

krrong replied:
이 부분은 매번 놓치는 것 같네요.. 감사합니다!🤠

sujin9 commented on app/src/main/java/woowacourse/paint/MainActivity.kt:
```diff
@@ -1,14 +1,51 @@
-class MainActivity : AppCompatActivity() {
+class MainActivity : AppCompatActivity(), OnClickListener {
+    private lateinit var viewBinding: ActivityMainBinding
+    private var thickness = 0f
+    private var color = 0
```
변수명이 모호하다는 생각이 들어요 🤔 
`MainActivity`에서는 그림판이라는 기능이 잘 나타나지 않기 때문에
무엇의 두께이고 무엇의 색상인지, 
해당 변수들이 왜 어디서 필요한지 잘 모를 수 있을 것 같아요.

krrong replied:
변수명을 조금 더 직관적으로 변경하여 반영했습니다!🙃

sujin9 reviewed (changes requested):
<image src="https://github.com/woowacourse/android-paint/assets/8490058/cc88c9e5-2a78-4909-9797-bfd9bfdc499b" width=360>

안녕하세요 크롱! 🦖 
만나게 되어 반갑습니다. 잘 부탁드려요 ~~
(반가운 마음을 담아 마우스로 한땀 한땀 그려보았습니다. 😎)

이미 말씀을 드렸었지만, 제가 먼저 미션을 선행해야 더 많은 리뷰를 수월하게 드릴 수 있을 것 같아
리뷰가 늦어지게 되었네요 😢 상황 이해해주셔서 감사해요! ;)

전체적으로 여러 경우의 수를 고려하고 고민해서 구현해주신 것 같아요. 👍 
기존에 예상한 구현 방식과 많이 달라서 재밌었어요. ;)

리뷰에 몇 가지 고려해볼 점이나 제안 사항들을 적어보았으니 확인해주세요!
혹시 이해가 가지 않으시거나 궁금한 점이 있으시면 댓글이나 DM 주세요! 😊

sujin9 commented on app/src/main/res/layout/activity_main.xml:
```diff
@@ -6,4 +6,86 @@
 
+    <woowacourse.paint.custom.CustomView
+        android:id="@+id/custom_canvas"
+        android:layout_width="match_parent"
+        android:layout_height="match_parent" />
```
한 가지가 누락됐네요..!
앱을 실행해보니 브러시 색상 및 두께를 조절하는 도구 패널 뒷부분에도 그림이 그려지더라구요.
혹시 의도하신 부분일까요?? 😮 (정말 궁금해서 여쭤봅니다! ㅋㅋ)

krrong replied:
넵 정확히 의도한 것입니다~  
하지만 RecyclerView로 변경하면서 도구 패널 뒤에는 그리지 못하게 되었답니다...😢

krrong committed:
refactor: 클래스명 변경 및 패키지 구조 변경

krrong committed:
feat: 리사이클러뷰 구성요소 추가

krrong committed:
refactor: 리사이클러뷰로 변경

krrong committed:
refactor: RangeSlider 설정 xml로 이동

krrong committed:
refactor: index를 0부터 시작하도록 변경

krrong committed:
refactor: match_parent에서 0dp로 변경

krrong committed:
refactor: 그림 그리는 방식 변경

- 점을 잇는 방식에서 lineTo 함수 사용
- 점을 관리하는 방식(Point)에서 라인을 관리(CurveLine)하는 방식으로 변경

krrong committed:
refactor: CurveLine을 관리하는 일급 컬렉션 생성

krrong committed:
refactor: CurveLine 추가하는 위치 변경

- ACTION_MOVE 에서 ACTION_DOWN으로 변경
- ACTION_MOVE 에서 할 경우 선분을 몇개 그리지 않아도 굉장한 성능 저하 발생(여러개의 선분이 추가되기 때문)

krrong committed:
refactor: lineTo에서 quadTo로 변경

krrong committed:
refactor: 변수명과 함수명을 조금 더 직관적으로 변경

krrong committed:
feat: 도메인 모듈 추가

krrong committed:
feat: Brush 도메인 클래스 생성 및 테스트

krrong committed:
refactor: 접근 제한자 변경 private -> public

krrong committed:
feat: domain 모듈 의존성 추가

krrong committed:
refactor: BrushWidth 자료형을 Float으로 변경

krrong committed:
feat: brush 관련 ui model 생성

krrong committed:
refactor: 패키지 구조 변경

krrong committed:
refactor: 색상 리스트 생성방법 변경

krrong committed:
feat: viewModel 의존성 추가

krrong committed:
refactor: BrushColorUiModel 적용

krrong committed:
refactor: viewModel 적용

krrong committed:
test: Brush Test 추가

krrong committed:
style: 코드 정리

krrong committed:
refactor: CurveLine의 속성 변경을 CurveLine 하도록 변경

krrong committed:
refactor: Activity 에서 Brush 속성을 가지고 있지 않도록 변경

krrong committed:
refactor: inflate를 ViewHolder에서 진행하도록 수정

krrong committed:
refactor: inflate를 ViewHolder에서 진행하도록 수정

krrong committed:
refactor: Path 추가를 CurveLine 하도록 변경

krrong committed:
refactor: inflate를 ViewHolder에서 진행하도록 수정

sujin9 commented on app/build.gradle.kts:
```diff
@@ -43,6 +43,10 @@ dependencies {
     implementation("androidx.appcompat:appcompat:1.6.0")
     implementation("com.google.android.material:material:1.7.0")
     implementation("androidx.constraintlayout:constraintlayout:2.1.4")
+    implementation(project(mapOf("path" to ":domain")))
```
도메인 모듈을 분리하신 이유가 무엇인가요? 정말 궁금해서 여쭤봅니다! 
더불어 domain model과 ui model을 나누신 이유도요!

sujin9 commented on app/src/main/java/woowacourse/paint/custom/view/CanvasView.kt:
```diff
@@ -0,0 +1,69 @@
+
+        when (event.action) {
+            MotionEvent.ACTION_DOWN -> {
+                startDrawing(pointX, pointY)
```
단순히 화면을 클릭하면 점이 찍히지 않아요.
곡선 뿐만 아니라 점도 찍을 수 있도록 개선해보면 어떨까요?

sujin9 commented on app/src/main/java/woowacourse/paint/custom/model/CurveLine.kt:
```diff
@@ -0,0 +1,46 @@
+        paint.strokeCap = Paint.Cap.ROUND
+    }
+
+    fun changeStrokeWidth(new: BrushWidth): CurveLine {
+        return this.copy(
+            path = Path(),
+            paint = Paint(this.paint).apply { strokeWidth = new.width },
+        )
+    }
+
+    fun changeColor(new: BrushColorUiModel): CurveLine {
+        return this.copy(
+            path = Path(),
+            paint = Paint(this.paint).apply { color = new.color },
+        )
+    }
```
꼼꼼하게 깊은 복사를 해주셨네요! 👍

sujin9 commented on app/src/main/java/woowacourse/paint/mapper/BrushColorMapper.kt:
```diff
@@ -0,0 +1,24 @@
+
+fun BrushColor.toBrushColorUiModel(): BrushColorUiModel {
+    return when (this) {
+        BrushColor.RED -> BrushColorUiModel.RED
+        BrushColor.YELLOW -> BrushColorUiModel.YELLOW
+        BrushColor.ORANGE -> BrushColorUiModel.ORANGE
+        BrushColor.GREEN -> BrushColorUiModel.GREEN
+        BrushColor.BLUE -> BrushColorUiModel.BLUE
```
이 경우도, 색상 종류가 많아진다면 어떻게 될까요? 🤔 (이하 동일)

위 선택한 코드 부분에 대해 예시를 들자면,
```suggestion
        return BrushColorUiModel.values().firstOrNull { it.name == this.name } ?: BrushColorUiModel.RED
```
이런 식으로 코드를 간결하게 표현해볼 수 있을 것 같아요!
꼭 name을 비교하는 방법이 아니어도 방법은 다양하게 있을 것 같네요!

다만, 제안드린 코드를 예시로 들면
name이 같지 않다거나 하는 실수가 발생할 수 있다는 위험성이 있을 것 같아요.
더 좋은 방법을 고민해보시면 좋을 것 같고, 크롱의 생각에 따라 반영 여부를 결정하시면 될 것 같네요! ;)

sujin9 commented on app/src/main/java/woowacourse/paint/palette/PaletteViewHolder.kt:
```diff
@@ -0,0 +1,23 @@
+    fun bind(brushColorUiModel: BrushColorUiModel, onPaletteClick: (BrushColorUiModel) -> Unit) {
+        binding.palette.apply {
+            setBackgroundColor(brushColorUiModel.color)
+            setOnClickListener {
+                onPaletteClick(brushColorUiModel)
+            }
```
`onClickListener`를 bind할 때마다 등록해줄 필요가 있을까요?

sujin9 commented on app/src/main/java/woowacourse/paint/palette/PaletteViewHolder.kt:
```diff
@@ -0,0 +1,23 @@
+package woowacourse.paint.palette
```
이건 여담인데요..
MainActivity 내의 컬러 팔레트에 대한 Adapter/ViewHolder라고 생각했는데
main패키지에도 없고 presentation 패키지에도 없어서
파일을 찾는 게 조금 헷갈렸어요 😅 

혹시 크롱의 패키지 관리 기준이 따로 있으신걸까요?

sujin9 commented on app/src/main/java/woowacourse/paint/custom/model/CurveLine.kt:
```diff
@@ -0,0 +1,46 @@
+        lastY = y
+    }
+
+    fun quadTo(x: Float, y: Float) {
+        path.quadTo(lastX, lastY, (x + lastX) / 2, (y + lastY) / 2)
+        lastX = x
+        lastY = y
+    }
```
quadTo 사용!! 👍

sujin9 commented on app/src/main/java/woowacourse/paint/presentation/uimodel/BrushColorUiModel.kt:
```diff
@@ -0,0 +1,11 @@
+import androidx.annotation.ColorInt
+
+enum class BrushColorUiModel(@ColorInt val color: Int) {
+    RED(0xFFEA3323.toInt()),
+    YELLOW(0xFFFCEC60.toInt()),
+    ORANGE(0xFFF19D38.toInt()),
+    GREEN(0xFF67AD5B.toInt()),
+    BLUE(0xFF4350AF.toInt()),
```
`colors.xml`에 있던 리소스를 삭제하고 이렇게 직접 적어주신 이유가 궁금해요! 🧐

sujin9 commented on app/src/main/java/woowacourse/paint/custom/view/CanvasView.kt:
```diff
@@ -0,0 +1,69 @@
+            }
+            else -> super.onTouchEvent(event)
+        }
+        return true
```
```suggestion
        invalidate()
        return true
```
아주 사소한 거긴 한데요..
중복되는 코드를 이렇게 빼내올 수 있을 것 같네요 ~

sujin9 commented on app/src/main/res/layout/activity_main.xml:
```diff
@@ -1,9 +1,42 @@
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent" />
+
+    <com.google.android.material.slider.RangeSlider
```
`RangeSlider` 말고 `Slider`도 있어요!
![image](https://github.com/woowacourse/android-paint/assets/8490058/d8c74c18-513f-4b1a-823e-a71261b78f83)

저는 개인적으로 Slider가 더 적합하다고 생각해서 변경했거든요..!
(`범위`를 설정할 필요가 없다고 생각했고, Slider의 `android:value` 속성을 통해 default값을 정해줄 수 있기 때문입니다)

크롱에게도 이런 게 있다~ 하고 알려드리고 싶어서 소개해요 :)
Slider와 RangeSlider 중 원하는 걸로 사용하시면 됩니다 ~

ref. 
https://m3.material.io/components/sliders/guidelines

sujin9 commented on domain/src/test/java/com/now/domain/BrushTest.kt:
```diff
@@ -0,0 +1,43 @@
+
+        // when & then
+        assertEquals(brush.brushColor, BrushColor.RED)
+        assertEquals(brush.brushWidth, BrushWidth(1f))
```
크롱의 그림판은 브러시 두께가 5단위로 나뉘어져 있는데,
Default값도 5단위로 맞춰주시는 건 어떨까요? 작은 제안 드립니다~ :)

sujin9 reviewed (approved):
안녕하세요 크롱 ! 🦖 

연휴동안 엄청 열심히 하신 것 같아요!
코드 구성과 구현 방법이 완전히 달라졌던데요!?
새롭게 리뷰하는 기분이었습니다. ㅋㅋㅋㅋ
모델을 잘게 나누고, 테스트까지 꼼꼼하게 해주신 게 인상적이었어요.

다음 미션 진행을 위해 이번 미션은 머지하겠습니다!
남겨드린 코멘트들은 생각해보시고 다음 스텝 때 이어서 얘기하면 좋을 것 같아요 :)

고생 많으셨습니다 ~
